src/BucketTesting/Validation/CampaignUtilizationValidator.php

```diff
@@ -0,0 +1,137 @@
+		if ( $this->hasValidated ) {
+			return;
+		}
+
+		$featureToggleChecks = $this->getFeatureToggleChecks();
+		if ( $duplicates = $this->getDuplicates( $featureToggleChecks ) ) {
+			foreach ( $duplicates as $duplicate ) {
+				$this->errorLogger->addError(
+					'ChoiceFactory contains multiple checks for feature toggle: ' . $duplicate
+				);
+			}
+			return;
+		}
+
+		$configBucketIdList = $this->buildBucketIdsFromCampaignCollection(
+			$this->campaignCollection
+		);
+
+		/** Remove all buckets which were marked as test buckets in TEST_BUCKET_IGNORE_LIST */
+		$featureToggleChecks = array_filter( $featureToggleChecks, [ __CLASS__, 'filterBucketIds' ] );
+		$configBucketIdList = array_filter( $configBucketIdList, [ __CLASS__, 'filterBucketIds' ] );
+
+		/** Check if either the configuration files or the choice factory has rules not found in the other */
+		$codeConfigDiff = array_diff( $featureToggleChecks, $configBucketIdList );
+		$configCodeDiff = array_diff( $configBucketIdList, $featureToggleChecks );
+
+		if ( empty( $codeConfigDiff ) === false || empty( $configCodeDiff ) === false ) {
+			foreach ( $codeConfigDiff as $missingBucket ) {
+				$this->errorLogger->addError(
+					'Bucket ' . $missingBucket . ' is configured but no implementation can be found in ChoiceFactory.'
+				);
+			}
+			foreach ( $configCodeDiff as $featureToggleCheck ) {
+				$this->errorLogger->addError(
+					'Feature toggle check for ' . $featureToggleCheck . ' is implemented but no campaign configuration can be found.'
+				);
+			}
+		}
+
+		$this->hasValidated = true;
+	}
+
+	private function getFeatureToggleChecks(): array {
+		$featureToggleChecks = [];
+		foreach ( $this->parseMethodCalls() as $featureToggleCheck ) {
+			if ( count( $featureToggleCheck->args ) !== 1 ) {
+				throw new \LogicException( self::FEATURE_TOGGLE_METHOD_NAME . ' should have exactly one argument.' );
+			}
+			$featureToggleChecks[] = $featureToggleCheck->args[0]->value->value;
+		}
+		return $featureToggleChecks;
+	}
+
+	private function parseMethodCalls(): array {
+		$parser = ( new ParserFactory() )->create( ParserFactory::PREFER_PHP7 );
+		$nodeFinder = new NodeFinder();
+		$choiceFactoryCode = file_get_contents( self::CHOICE_FACTORY_LOCATION );
+		if ( $choiceFactoryCode === false || $choiceFactoryCode === '' ) {
+			throw new FileNotFoundException( null, 0, null, self::CHOICE_FACTORY_LOCATION );
+		}
+		$syntaxTree = $parser->parse( $choiceFactoryCode );
+		return $nodeFinder->find(
+			$syntaxTree,
+			function ( Node $node ) {
+				return $node instanceof MethodCall && $node->name->toString() === self::FEATURE_TOGGLE_METHOD_NAME;
+			}
+		);
+	}
+
+	private function buildBucketIdsFromCampaignCollection( CampaignCollection $campaignCollection ): array {
+		$buckets = [];
+		/** @var Campaign $campaign */
+		foreach ( $campaignCollection as $campaign ) {
+			foreach ( $campaign->getBuckets() as $bucket ) {
+				$buckets[] = 'campaigns.' . $bucket->getId();
```

Comment:
This is the second place we the `campaigns` string comes out of the blue (the other, related place is `CampaignFeatureBuilder::getFeatureName`). I suggest we add the `campaigns` prefix in `Bucket::getId` and call that function in both places.